### PR TITLE
Let `css` prop attachment be distributed over union types.

### LINKED ISSUE #3185: The `css` prop is not attached to a type of props if it is a disjoint union type on `className` field.
**Current behavior:**

`WithConditionalCSSProp<P>` doesn't create the `css` prop if `P` is a disjoin union type on `className` field. Therefore we can't use some components optionally taking `className` field, like [`Link` of wouter](https://github.com/molefrog/wouter/blob/607d2051c5ed9c3edd680c3e8ece75bf95a0b4fd/packages/wouter/types/index.d.ts#L120).

**To reproduce:**

Since it's not a problem inherent in emotion or react but a problem of treating union types, we can reproduce it with following simplified snippet.

```typescript
type WithConditionalCSSProp<P> = 'className' extends keyof P
  ? string extends P['className' & keyof P]
    ? { css?: 'Interpolation<Theme>' } // To make the snippet simpler, I've just used string literal instead.
    : {}
  : {}

type BUG = WithConditionalCSSProp<{ className: string, foo: number } | { bar: boolean }> 
```
[typescript playground](https://www.typescriptlang.org/play/?#code/C4TwDgpgBA6glsAFgYQPYDsAmC4YIYA2yAysQAoBOqYAPGQHxQC8UA5AMYF4DO3AcngC2EVlAgAPYBCzcoAawghUAMyhkAUFCgB+KN2AU46AOZjJ0zLLIBtDl14DhogGTzFKtQF1NWnVADeUOy82gBcbACS6FIUYKhcwLjoNAAqiBDC9KIAvj5a4f65+QG56qCQUABCAKoA4sywCCgY2In4RKSU1DSBnDz8QhDh+oYmADRQyqio4egAroIARhAUUNlQAD4BUIt4FOGL0wQQeOhrjEA)

However, It's always good to have detailed reproduction. Here is [the case I encountered](https://codesandbox.io/p/sandbox/emotion-issue-template-forked-sz4v4c?file=%2Fsrc%2FApp.tsx%3A10%2C1&layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clw6m2x1c00063b5t9tr5bkwh%2522%252C%2522sizes%2522%253A%255B100%252C0%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clw6m2x1c00023b5t5skevgvz%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clw6m2x1c00033b5tkwoica79%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clw6m2x1c00053b5thpmvjdcs%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clw6m2x1c00023b5t5skevgvz%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clw6m2x1b00013b5treq9nemo%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fsrc%252Findex.tsx%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522id%2522%253A%2522clw6mfd6b00023b5t8vhl6636%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A10%252C%2522startColumn%2522%253A1%252C%2522endLineNumber%2522%253A10%252C%2522endColumn%2522%253A1%257D%255D%252C%2522filepath%2522%253A%2522%252Fsrc%252FApp.tsx%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%252C%2522id%2522%253A%2522clw6m2x1c00023b5t5skevgvz%2522%252C%2522activeTabId%2522%253A%2522clw6mfd6b00023b5t8vhl6636%2522%257D%252C%2522clw6m2x1c00053b5thpmvjdcs%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clw6m2x1c00043b5tvazy1b7a%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A0%252C%2522path%2522%253A%2522%252F%2522%257D%255D%252C%2522id%2522%253A%2522clw6m2x1c00053b5thpmvjdcs%2522%252C%2522activeTabId%2522%253A%2522clw6m2x1c00043b5tvazy1b7a%2522%257D%252C%2522clw6m2x1c00033b5tkwoica79%2522%253A%257B%2522tabs%2522%253A%255B%255D%252C%2522id%2522%253A%2522clw6m2x1c00033b5tkwoica79%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Afalse%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D).

image

**Expected behavior:**

I think it might be more accurate to add `css` prop to branches where proper `className` field is available. Therefore pass type check for components optionally taking `className` like wouter's.

I suggest following as a fix:

```typescript
// Make LibraryManagedAttributes distributed over unions
type LibraryManagedAttributes<C, P> = P extends unknown ? 
  WithConditionalCSSProp<P> & ReactJSXLibraryManagedAttributes<C, P> 
  : never
```

**Environment information:**

- `react` version: 18.2.0
- `@emotion/react` version: 11.11.4

### PR DESCRIPTION
**Pull request by ENvironmentSet**
**What**: Let `css` prop attachment be distributed over union types. So that branches with `className` prop are extended with `css` prop.

**Why**: Some components may accept `className` prop conditionally. For example, [`Link` of wouter does.](https://github.com/molefrog/wouter/blob/f953ae6811c4e49201e11844038b71e531ab76d2/packages/wouter/types/index.d.ts#L129) Emotion skips these cases, however, and does not add `css` prop in anywhere. As far as I know, there is no runtime or compilation-time problem stopping us from supporting these kind of components. Thus I propose this change to enhance expressiveness of the library.

fixes #3185

**How**: Made `LibraryManagedAttributes` be distributed over a prop type in cases it is a union.

**Checklist**:

- [ ] Documentation "N/A"
- [x] Tests
- [x] Code complete
- [x] Changeset 

Should I add test cases for this change? Please let me know if you need.

### CODE REVIEW

**Review by Andarist — CHANGES_REQUESTED** [review]
We'd need some type-level tests to take in this change

**Comment by ENvironmentSet** [comment]
Indeed, I'll add some. Would you please give me a guideline to add type level tests or example to refer to? I could not find either an example of type-level test or a setting relevant to it in emotion repository.

**Comment by Andarist** [comment]
You can add them here at the bottom of the file: https://github.com/emotion-js/emotion/blob/e819f9c39c4485997ce97fd2cfd330be8114e0f8/packages/react/types/tests.tsx

**Comment by ENvironmentSet** [comment]
Thanks, I added a test. Let me know if you need more.

**Comment by ENvironmentSet** [comment]
It seems dtslint compares type too strictly. I'll find other ways to express validity of the implementation.

File: packages/react/types/tests.tsx [hunk 1]
```diff
@@ -6,7 +6,9 @@ import {
   css,
   jsx,
   keyframes,
-  withEmotionCache
+  withEmotionCache,
+  Interpolation,
+  Theme
```
Comment (ENvironmentSet) [inline packages/react/types/tests.tsx:11]: Oops, sorry. I forgot to delete unused imports. I’ll add additional commit as soon as possible.

File: packages/react/types/tests.tsx [hunk 2]
```diff
@@ -240,6 +240,18 @@ const anim1 = keyframes`
...
+     Additional brackets here prevent type reference('EmotionJSX.LibraryManagedAttributes<...>') from being resolved to type alias name('_HasCssPropAsIntended7').
+     Without them, dtslint will compare the expected type to the type name.
+  */
+  type _HasCssPropAsIntended7 = [
+    // $ExpectType { foo: boolean; } | ({ css?: Interpolation<Theme>; } & { foo: number; className: string; })
+    EmotionJSX.LibraryManagedAttributes<
+      {},
+      { foo: number; className: string } | { foo: boolean }
+    >
+  ]
```
Comment (Andarist) [inline packages/react/types/tests.tsx:253]: This is like a unit test. I am more interested in an integration test for this change. How did you come across this to be an issue? Maybe it was related to code like this?
```ts
type DiscriminatedClassNameSupportProp =
  | {
      variant: "foo";
      className?: string;
    }
  | {
      variant: "bar";
    };

function DiscriminatedClassNameSupportComponent(
  props: DiscriminatedClassNameSupportProp,
) {
  return null;
}

<DiscriminatedClassNameSupportComponent
  variant="foo"
  // this should be OK
  css={{ color: "hotpink" }}
/>;

<DiscriminatedClassNameSupportComponent
  variant="bar"
  // this shouldn't be OK
  css={{ color: "hotpink" }}
/>;
```

Comment (ENvironmentSet) [reply packages/react/types/tests.tsx:253]: Yes, that was my case. May I add tests like that?

Comment (Andarist) [reply packages/react/types/tests.tsx:253]: Yes. Please replace the tests you have here with something closer to the one above :)

Comment (ENvironmentSet) [reply packages/react/types/tests.tsx:253]: Sure. I'll update PR tonight. Thank you for comments!

**Pull request merged**